compute: tokenize error logging

teskje opened the pull request:
This PR modifies the error logging behavior of compute operators to only log messages to Sentry when the dataflow is not in the process of shutting down. This avoids producing noise caused by dataflow operators emitting partial update sets when they are cancelled. It is a precondition for enabling the faster dataflow shutdown methods described in the ["Faster Dataflow Shutdown" design](https://github.com/MaterializeInc/materialize/pull/18760).

The shutdown checking is performed through a shutdown token that is wrapped inside a new `ErrorLogger` type that gets passed around to the various operator renderers. An `ErrorLogger` encapsulates both shutdown checking and our `warn!`/`error!` logging format.

### Motivation

  * This PR adds a known-desirable feature.

Enables https://github.com/MaterializeInc/materialize/pull/18760.

### Tips for reviewer

Look at the commits and their descriptions separately!

* The first commit moves the `MaybeValidatingRow` trait into a new `mz_compute::render::errors` module, which is also where the new `ErrorLogger` type lives.
* The second commit adds the `ErrorLogger` type.
* The third commit attaches a shutdown token to the rendering `Context` and updates the top-level rendering code to make sure the token gets added to the exports depending on it.
* The fourth commits adjusts error logging in dataflow operators to use the `ErrorLogger`, rather than using the `warn!`/`error!` macros directly.
* The fifth commit makes most of the free-standing reduce and top-k rendering functions methods of the `Context` type. Rather than requiring `debug_name` and `error_logger` arguments, they can now get these fields directly from the context, reducing the number of function arguments that need to be passed.

Unfortunately, the diff for the last commit is quite hairy, since it changes not only the indentation level, but also causes rustfmt to re-wrap many of the higher-indented lines. Hiding whitespace changes helps a bit, but still produces a diff much larger than it should be. Since the last commit is a pure code organization change and doesn't affect functionality, I would be fine with dropping it if you'd rather not review it.

~~One thought about the `ErrorLogger` type: It would be quite easy to also wrap the `debug_name`, in addition to the token, inside it. Then we wouldn't have to pass that around in rendering anymore. The drawback is that we'd lose our current flexibility in formatting the "details" part. We'd probably need to put the dataflow name as a fixed prefix of the message, so constructs like "found invalid accumulations in dataflow {debug_name}" would no longer be possible.~~ This is now implemented in the sixth commit!

If anyone has an idea how to actually test this change, I'd be happy to hear it! I couldn't come up with a way to provoke invalid retractions during dataflow shutdown (they should be impossible today), much less doing so in a reliable way.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

vmarcos commented on src/compute/src/render/reduce.rs:
```diff
@@ -387,15 +412,15 @@ where
+                             n_aggregates_requested={requested}, \
+                             n_distinct_aggregate_types={n_distinct_aggregate_types}",
+                            requested = input.len(),
+                        ),
```
This fragment, and others similar to it, seem to change the previous code to omit `message` from what is written to the `warn!` output. Is this intentional? I would find it a bit easier if `message` were included in the warning as well, so that it is easier to match up in the output logs; this is just personal preference, of course.

vmarcos replied:
I think the message in the warning adds some value to sort out where these breadcrumbs are coming from without having to refer directly to the code. We may also be getting some of these outputs from multiple operators if they are all seeing invalid accumulations at once.

teskje replied:
The `ErrorLogger` now always logs the `message` in the WARN log as well, to enable this matching. This is also true for instances (mostly calls to `soft_panic_or_log`) where this previously was not the case (i.e. the ERROR message was different from the WARN message).

tokenrove commented on src/compute/src/render/reduce.rs:
```diff
@@ -508,8 +541,12 @@ where
                         continue;
                     }
                     let message = "Non-positive multiplicity in DistinctBy Retractions";
-                    warn!(?row, ?count, debug_name, "[customer-data] {message}");
```
One downside to this approach is that we lose structured field logging.  How ugly would it be to supply some macros that take an `ErrorLogger`?

teskje replied:
I've not been able to come up with anything ergonomic so far. We could probably introduce something like:

```rust
log_dataflow_error!(
    self.error_logger(),
    "Non-positive multiplicity in DistinctBy Retractions",
    ?row, ?count, debug_name,
    "[customer-data] Non-positive multiplicity in DistinctBy Retractions",
);
```

Doesn't seem too bad, actually. Especially because we could make it possible to reference the message in the details format string, and make the `debug_name` be implicitly added to all logs.

```rust
log_dataflow_error!(
    self.error_logger(),
    "Non-positive multiplicity in DistinctBy Retractions",
    ?row, ?count,
    "[customer-data] {message}",
);
```

I'm just not sure whether my macro foo is up to this task :)

In any case, I'd rather make this the scope of another PR, since I foresee some bikeshedding.

tokenrove replied:
Nice!   I'm also happy to do this if you don't want to deal with it.

vmarcos reviewed (commented):
The first four commits look good to me; I only have a minor remark about some warning messages that appear to have been shortened in some places.

I am uncertain about the last commit. The previous code organization was already in place prior to my start in Materialize and I do not think I have a clear understanding of why these functions were not part of `Context` to start with or otherwise why they were not collected under some other abstraction. Perhaps we can ask for someone with historical background to manifest before moving on with this change? Also, I did not give enough thinking to the question of whether there may also be a better code organization that we could think of rather than having all of the functions collected under `Context`.

tokenrove commented on src/compute/src/render/errors.rs:
```diff
@@ -70,3 +70,48 @@ where
+    ///    We require it to be a static string, to ensure that Sentry always merges instances of
+    ///    the same error together.
+    ///  * `details` is logged at WARN level and will appear in the breadcrumbs.
+    ///    Put relevant dynamic information here.
```
Along the same lines, there could be variants here like `log_customer_data` or `log_with_sensitive_details` that adds the `[customer-data]` string to the warn itself.

tokenrove replied:
> What do you think of just adding the tag even with the regular `log` method? We probably always want to log the `debug_name`, which was picked by the customer. Do we anticipate cases where the details don't contain customer data?

I think it's better to err on the side of assuming any fields we log in this code are sensitive.

teskje replied:
Details are now always logged with the `[customer-data]` tag and I've added a TODO to make it clear that finalizing error handling strategy is tracked by #18214.

tokenrove commented on src/compute/src/render/context.rs:
```diff
@@ -114,6 +121,7 @@ where
             as_of_frontier,
             until: dataflow.until.clone(),
             bindings: BTreeMap::new(),
+            shutdown_token: Weak::new(),
```
There's never a case where we create a context like this and don't populate `shutdown_token`, right?  Because it seems like if that happened, the error logging would not log in those cases.  The code currently looks fine, since I guess there must always be at least one object to build, but I'm just wondering if there's any way to enforce that you can't accidentally create an error-muffling context.

teskje replied:
Yeah, this seems like a potential future pitfall. I'll make this field an `Option<Weak<()>>` instead.

teskje commented:
> I am uncertain about the last commit. The previous code organization was already in place prior to my start in Materialize and I do not think I have a clear understanding of why these functions were not part of Context to start with or otherwise why they were not collected under some other abstraction. Perhaps we can ask for someone with historical background to manifest before moving on with this change?

I asked [in Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1681737875995369) and there doesn't seem to be a specific reason.

This should be ready for another look! I've addressed the comments and added another commit that pulls the `debug_name` printing into the `ErrorLogger` as well, making the callsites a bit more DRY in the process.

vmarcos reviewed (approved):
Thanks for asking in Slack; the changes LGTM! I think it is quite tricky to test this at the moment, but maybe we can think about an approach when we allow `persist_source` to stop emitting early as well.

teskje commented:
TFTRs despite the huge diff!
